From a1b9d107965df0275913290e08af08a097eff8bc Mon Sep 17 00:00:00 2001
From: jgv <jgv@opencascade.com>
Date: Fri, 1 Feb 2013 17:22:26 +0400
Subject: [PATCH] [OCCT] 0023436: BRepOffsetAPI_ThruSections crashes when
 lofting through identical profiles Adding of test case

  Cherry-picked from
    http://git.dev.opencascade.org/gitweb/?p=occt.git;a=commit;h=aadab51
---
 src/BRepOffsetAPI/BRepOffsetAPI_ThruSections.cxx |   22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

Index: oce/src/BRepOffsetAPI/BRepOffsetAPI_ThruSections.cxx
===================================================================
--- oce.orig/src/BRepOffsetAPI/BRepOffsetAPI_ThruSections.cxx
+++ oce/src/BRepOffsetAPI/BRepOffsetAPI_ThruSections.cxx
@@ -102,7 +102,27 @@
 #include <BRepBuilderAPI_FindPlane.hxx>
 
 
+//=======================================================================
+//function : PreciseUpar
+//purpose  : pins the u-parameter of surface close to U-knot
+//           to this U-knot
+//=======================================================================
+
+static Standard_Real PreciseUpar(const Standard_Real anUpar,
+                                 const Handle(Geom_BSplineSurface)& aSurface)
+{
+  Standard_Real Tol = Precision::PConfusion();
+  Standard_Integer i1, i2;
 
+  aSurface->LocateU(anUpar, Tol, i1, i2);
+  Standard_Real U1 = aSurface->UKnot(i1);
+  Standard_Real U2 = aSurface->UKnot(i2);
+
+  Standard_Real NewU = anUpar;
+
+  NewU = (anUpar - U1 < U2 - anUpar)? U1 : U2;
+  return NewU;
+}
 
 //=======================================================================
 //function :  PerformPlan
@@ -629,6 +649,8 @@ void BRepOffsetAPI_ThruSections::CreateS
     Standard_Real Ui1,Ui2,V0,V1;
     Ui1 = i-1;
     Ui2 = i;
+    Ui1 = PreciseUpar(Ui1, surface);
+    Ui2 = PreciseUpar(Ui2, surface);
     V0  = surface->VKnot(surface->FirstVKnotIndex());
     V1  = surface->VKnot(surface->LastVKnotIndex());
     surface->Segment(Ui1,Ui2,V0,V1);
